ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+		observers:          make(map[observable.Observer]struct{}),
+		lm:                 lifecycle.NewLifecycleManager(),
+		checkpointDistance: checkpointDistance,
+		checkpointsToKeep:  checkpointsToKeep,
+	}, nil
+}
+
+func (c *Compactor) Subscribe(observer observable.Observer) {
+	var void struct{}
+	c.observers[observer] = void
+}
+
+func (c *Compactor) Unsubscribe(observer observable.Observer) {
+	delete(c.observers, observer)
+}
+
+// Ready periodically fires Run function, every `interval`
+func (c *Compactor) Ready() <-chan struct{} {
+	c.lm.OnStart(func() {
+		go c.run()
+	})
+	return c.lm.Started()
+}
+
+func (c *Compactor) Done() <-chan struct{} {
+	c.lm.OnStop(func() {
+		// Notify observers
+		for observer := range c.observers {
+			observer.OnComplete()
+		}
+
+		// Signal Compactor goroutine to stop
+		doneCh := make(chan struct{})
+		c.stopCh <- doneCh
+
+		// Wait for Compactor goroutine to stop
+		<-doneCh
+
+		// Close trieUpdateDoneCh to signal trie updates are finished
+		close(c.trieUpdateDoneCh)
+	})
+	return c.lm.Stopped()
+}
+
+func (c *Compactor) run() {
+
+	// checkpointSem is used to limit checkpointing to one.
+	// If previous checkpointing isn't finished when enough segments
+	// are finalized for next checkpointing, retry checkpointing
+	// again when next segment is finalized.
+	// This avoids having more tries in memory than needed.
+	checkpointSem := semaphore.NewWeighted(1)
+
+	checkpointResultCh := make(chan checkpointResult, 1)
+
+	// Get active segment number.
+	// activeSegmentNum is updated when record is written to a new segment.
+	_, activeSegmentNum, err := c.wal.Segments()
+	if err != nil {
+		c.logger.Error().Err(err).Msg("compactor failed to get active segment number")
+		activeSegmentNum = -1
+	}
+
+	lastCheckpointNum, err := c.checkpointer.LatestCheckpoint()
+	if err != nil {
+		c.logger.Error().Err(err).Msg("compactor failed to get last checkpoint number")
+		lastCheckpointNum = -1
+	}
+
+	// Compute next checkpoint number.
+	// nextCheckpointNum is updated when
+	// - checkpointing starts, fails to start, or fails.
+	// - tries snapshot fails.
+	// NOTE: next checkpoint number must >= active segment num.
+	// We can't reuse mtrie state to checkpoint tries in older segments.
+	nextCheckpointNum := lastCheckpointNum + int(c.checkpointDistance)
+	if activeSegmentNum > nextCheckpointNum {
+		nextCheckpointNum = activeSegmentNum
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+Loop:
+	for {
+		select {
+
+		case doneCh := <-c.stopCh:
+			defer close(doneCh)
+			cancel()
+			break Loop
+
+		case checkpointResult := <-checkpointResultCh:
+			if checkpointResult.err != nil {
+				c.logger.Error().Err(checkpointResult.err).Msgf(
+					"compactor failed to checkpoint %d", checkpointResult.num,
+				)
+
+				// Retry checkpointing after active segment is finalized.
+				nextCheckpointNum = activeSegmentNum
+			}
+
+		case update, ok := <-c.trieUpdateCh:
```

Comment:
I think it's better to move the checkpoint related logic to a separate module, rather than mixing together here.

Basically, there are two pipelines:
The first pipeline, subscribes from trieUpdateCh and decide whether a segment has been created. Once created, it passes the segment to the next pipeline
The second pipeline, will subscribe the segment events, and decide whether to create a checkpoint. Since checkpointing takes time, if won't block the first pipeline from processing the trieUpdates, or passing the segment even if the checkpointing hasn't been done yet. 

Correct me if wrong. The problem of the current implementation is that during checkpointing, we won't process trie update.

Reply:
> Correct me if wrong. The problem of the current implementation is that during checkpointing, we won't process trie update.

That is incorrect.  Checkpointing is done in a [new goroutine.](https://github.com/onflow/flow-go/blob/65f20620dc38ec35efce75e407a6ee241410c091/ledger/complete/compactor.go#L201-L205)

Compactor can process trie update while checkpointing is running (they execute in parallel).  Also, semaphore makes sure only one checkpointing is running at a time.

Reply:
> Basically, there are two pipelines:
The first pipeline, subscribes from trieUpdateCh and decide whether a segment has been created. Once created, it passes the segment to the next pipeline
The second pipeline, will subscribe the segment events, and decide whether to create a checkpoint. Since checkpointing takes time, if won't block the first pipeline from processing the trieUpdates, or passing the segment even if the checkpointing hasn't been done yet.

Although I agree that there are two pipelines, I think they need to have different responsibilities from what you mentioned.  The goal is to not hold on to tries in memory that don't need to be checkpointed.

The first pipeline: subscribes from trieUpdateCh, decides whether a segment has been created, and decides whether a checkpoint needs to be created.  More importantly, it needs to take a snapshot of main ledger state at segment boundary when checkpointing is needed.  Taking a snapshot requires communication with `Ledger.Set` to ensure that WAL update and ledger state update are in sync.

The second pipeline is to create checkpoint file with ledger state snapshot taken from previous pipeline.  Checkpointing is done in a separate goroutine so it doesn't block the first pipeline.

Thoughts?